File: flutter_app/lib/models/recipe.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'recipe.freezed.dart';
part 'recipe.g.dart';

enum RecipeSource {
  @JsonValue('youtube')
  youtube,
  @JsonValue('instagram')
  instagram,
  @JsonValue('twitter')
  twitter,
  @JsonValue('cookpad')
  cookpad,
}

enum RecipeCategory {
  @JsonValue('肉料理')
  meat,
  @JsonValue('魚介料理')
  seafood,
  @JsonValue('野菜料理')
  vegetable,
  @JsonValue('サラダ')
  salad,
  @JsonValue('ご飯もの')
  rice,
  @JsonValue('麺類')
  noodle,
  @JsonValue('スープ・汁物')
  soup,
  @JsonValue('お菓子')
  sweets,
  @JsonValue('デザート')
  dessert,
  @JsonValue('パン')
  bread,
  @JsonValue('お弁当')
  bento,
  @JsonValue('その他')
  other,
}

extension RecipeCategoryExtension on RecipeCategory {
  String get displayName {
    switch (this) {
      case RecipeCategory.meat:
        return '肉料理';
      case RecipeCategory.seafood:
        return '魚介料理';
      case RecipeCategory.vegetable:
        return '野菜料理';
      case RecipeCategory.salad:
        return 'サラダ';
      case RecipeCategory.rice:
        return 'ご飯もの';
      case RecipeCategory.noodle:
        return '麺類';
      case RecipeCategory.soup:
        return 'スープ・汁物';
      case RecipeCategory.sweets:
        return 'お菓子';
      case RecipeCategory.dessert:
        return 'デザート';
      case RecipeCategory.bread:
        return 'パン';
      case RecipeCategory.bento:
        return 'お弁当';
      case RecipeCategory.other:
        return 'その他';
    }
  }
}

extension RecipeSourceExtension on RecipeSource {
  String get displayName {
    switch (this) {
      case RecipeSource.youtube:
        return 'YouTube';
      case RecipeSource.instagram:
        return 'Instagram';
      case RecipeSource.twitter:
        return 'Twitter';
      case RecipeSource.cookpad:
        return 'クックパッド';
    }
  }
}

@freezed
class Ingredient with _$Ingredient {
  const factory Ingredient({
    required String name,
    required String amount,
  }) = _Ingredient;

  factory Ingredient.fromJson(Map<String, dynamic> json) =>
      _$IngredientFromJson(json);
}

@freezed
class Recipe with _$Recipe {
  const factory Recipe({
    required int id,
    required String title,
    required String url,
    @JsonKey(name: 'image_url') required String imageUrl,
    String? notes,
    required RecipeSource source,
    List<String>? tags,
    @JsonKey(name: 'user_id') String? userId,
    List<Ingredient>? ingredients,
    @JsonKey(name: 'like_count') @Default(0) int likeCount,
    @JsonKey(name: 'is_liked_by_current_user') @Default(false) bool isLikedByCurrentUser,
  }) = _Recipe;

  factory Recipe.fromJson(Map<String, dynamic> json) => _$RecipeFromJson(json);
}

@freezed
class InsertRecipe with _$InsertRecipe {
  const factory InsertRecipe({
    required String title,
    required String url,
    @JsonKey(name: 'image_url') required String imageUrl,
    @JsonKey(includeIfNull: false) String? notes,
    required RecipeSource source,
    @JsonKey(includeIfNull: false) List<String>? tags,
    @JsonKey(name: 'user_id', includeIfNull: false) String? userId,
    @JsonKey(includeIfNull: false) List<Ingredient>? ingredients,
  }) = _InsertRecipe;

  factory InsertRecipe.fromJson(Map<String, dynamic> json) =>
      _$InsertRecipeFromJson(json);
}

File: nextjs-app/types/recipe.ts
export type RecipeSource = 'youtube' | 'instagram' | 'twitter' | 'cookpad'

export type RecipeCategory =
  | '肉料理'
  | '魚介料理'
  | '野菜料理'
  | 'サラダ'
  | 'ご飯もの'
  | '麺類'
  | 'スープ・汁物'
  | 'お菓子'
  | 'デザート'
  | 'パン'
  | 'お弁当'
  | 'その他'

export interface Ingredient {
  name: string
  amount: string
}

export interface Recipe {
  id: number
  title: string
  url: string
  image_url: string
  notes?: string
  source: RecipeSource
  tags?: string[]
  user_id?: string
  ingredients?: Ingredient[]
  like_count: number
  is_liked_by_current_user: boolean
  created_at: string
}

export interface InsertRecipe {
  title: string
  url: string
  image_url: string
  notes?: string
  source: RecipeSource
  tags?: string[]
  user_id?: string
  ingredients?: Ingredient[]
}

export interface UpdateRecipe {
  title?: string
  url?: string
  image_url?: string
  notes?: string
  source?: RecipeSource
  tags?: string[]
  ingredients?: Ingredient[]
}

export const RECIPE_SOURCES: { value: RecipeSource; label: string }[] = [
  { value: 'youtube', label: 'YouTube' },
  { value: 'instagram', label: 'Instagram' },
  { value: 'twitter', label: 'Twitter' },
  { value: 'cookpad', label: 'クックパッド' },
]

export const RECIPE_CATEGORIES: { value: RecipeCategory; label: string }[] = [
  { value: '肉料理', label: '肉料理' },
  { value: '魚介料理', label: '魚介料理' },
  { value: '野菜料理', label: '野菜料理' },
  { value: 'サラダ', label: 'サラダ' },
  { value: 'ご飯もの', label: 'ご飯もの' },
  { value: '麺類', label: '麺類' },
  { value: 'スープ・汁物', label: 'スープ・汁物' },
  { value: 'お菓子', label: 'お菓子' },
  { value: 'デザート', label: 'デザート' },
  { value: 'パン', label: 'パン' },
  { value: 'お弁当', label: 'お弁当' },
  { value: 'その他', label: 'その他' },
]
